### PR TITLE
Updated IPA Source Management

### PR DESCRIPTION
**Pull request by IsaacMarovitz**
https://user-images.githubusercontent.com/42140194/193436833-dc2819ca-d6fc-4d17-a657-5bae5b250d5a.mov

- Remove hardcoded `.json` both for flexability, and for copyright
- Allow users to add their own in-app IPA sources
- In-app downloads

**Other Changes/Bug Fixes:**
- `.app` folders will be renamed to match display name at install time. This means apps will now display with the correct name in the dock
- Updated the format of IPA Source `.json`s, an example of which can be found [here](https://raw.githubusercontent.com/IsaacMarovitz/PlayCover-IPA-Source-Tests/main/updatedsource.json)
- Fixed an issue where duplicate apps in the `Containers` folder with the same bundle ID would display incorrectly in grid view, but correctly in list view
- Fixed an issue where the display name of apps was sometimes equal to ""
- Cleaned up App Library and IPA Library views

### CODE REVIEW

File: PlayCover/AppInstaller/Installer.swift [hunk 1]
```diff
@@ -183,7 +183,11 @@ class Installer {
 
     /// Generates a wrapper bundle for an iOS app that allows it to be launched from Finder and other macOS UIs
     static func wrap(_ baseApp: BaseApp) throws -> URL {
-        let location = PlayTools.playCoverContainer.appendingPathComponent(baseApp.url.lastPathComponent)
+        let info = AppInfo(contentsOf: baseApp.url
+            .appendingPathComponent("Info.plist"))
+        let location = PlayTools.playCoverContainer
+            .appendingPathComponent(info.displayName)
+            .appendingPathExtension("app")
```
Comment (amirsaam) [inline PlayCover/AppInstaller/Installer.swift:190]: Does this address #304 ?

Comment (IsaacMarovitz) [reply PlayCover/AppInstaller/Installer.swift:190]: Maybe, but it wasn't written to fix that bug

Comment (amirsaam) [reply PlayCover/AppInstaller/Installer.swift:190]: We will find out then 👍

**Comment by amirsaam** [comment]
Shouldn't we have an `Add Source` button in the IPA Library View?
For better UX...

**Comment by IsaacMarovitz** [comment]
Mm good point, I've got a couple of UX improvements I can make

**Comment by amirsaam** [comment]
Since in-app download is implemented and we have IPA sources, does it check the IPA versions to notify newer versions available for update to user?

It can use notification function or an ⬆️ icon alongside apps that has been installed in IPA library

**Comment by Amachik** [comment]
can we merge this?

**Comment by IsaacMarovitz** [comment]
@amirsaam Currently it will take the most recent version in the IPA Source, it doesn't check the App Store currently

**Comment by IsaacMarovitz** [comment]
@Amachik All of our reviewers are currently busy so merges have been slow

**Comment by amirsaam** [comment]
> @amirsaam Currently it will take the most recent version in the IPA Source, it doesn't check the App Store currently

Why should it check App Store?
PlayCover should check App Library app version with the one from IPA Library that comes from IPA Source and tell there is difference in versions

File: PlayCover/Views/App Views/StoreAppView.swift [hunk 2]
```diff
@@ -109,31 +189,21 @@ struct StoreAppConditionalView: View {
             }
         }
         .task {
-            iconUrl = await getIconURLFromBundleIdentifier(app.id, app.region)
+            itunesData = await getITunesData(app.itunesLookup)
```
Comment (Depal1) [inline PlayCover/Views/App Views/StoreAppView.swift:192]: Leaving the "itunesLookup" attribute empty `""` for apps not available in the App Store for instance, results in a crash. This isn't desired behavior, as PlayCover should let you list apps that are not available in the App Store.

File: PlayCover/Views/Settings/IPASourceSettings.swift [hunk 3]
```diff
@@ -0,0 +1,263 @@
...
+
+    func validateSource(_ source: String) {
+        sourceValidationState = .checking
+        DispatchQueue.global(qos: .userInteractive).async {
+            if let url = URL(string: source) {
+                newSourceURL = url
+                if StoreVM.checkAvaliability(url: newSourceURL!) {
+                    do {
+                        if newSourceURL!.scheme == nil {
+                            newSourceURL = URL(string: "https://" + newSourceURL!.absoluteString)!
```
Comment (Depal1) [inline PlayCover/Views/Settings/IPASourceSettings.swift:233]: Local links are not being accepted as valid in the current scheme. `file:///path/to/my/apps/my_apps.json`

**Review by Depal1 — CHANGES_REQUESTED** [review]
Having app entries in the JSON with wrong information (like a mistyped link in `"itunesLookup"`) throws an error as it should, but if there are several non compliant entries there will be as many issue prompts, which in practice does not provide any more useful information to the user. Additionally, once the view is changed to grid, list, or from App Library, all the prompts are shown again in the same session.

When no internet connection is active, the issue describing `The internet connection appears to be offline` does not get triggered if all entries in the JSON files are compliant. On the other hand, when there are `n` non compliant app entries in a JSON, this issue is prompted `n` times.

**Comment by Depal1** [comment]
> > @amirsaam Currently it will take the most recent version in the IPA Source, it doesn't check the App Store currently
> 
> Why should it check App Store? PlayCover should check App Library app version with the one from IPA Library that comes from IPA Source and tell there is difference in versions

That is out of scope. PlayCover shouldn't monitor every app in a JSON list for updates from the App Store since it can't do anything meaningful about it (we can't fetch apps from the App Store), so it would amount to wasteful resource usage. Checking for updates is an exclusive task of the user.

**Comment by amirsaam** [comment]
@Depal1 hmmm, why would it check app store?
`ipa version in json >  ipa version installed = update available`

**Comment by Depal1** [comment]
> @Depal1 hmmm, why would it check app store?
> `ipa version in json > ipa version installed = update available`

Oh, okay. That is more sensible. Sure, it could be done.

File: PlayCover/Views/Settings/IPASourceSettings.swift (hunk 3 above)
Comment (IsaacMarovitz) [reply PlayCover/Views/Settings/IPASourceSettings.swift:233]: Do we really need to support local links?

**Review by Depal1 — APPROVED** [review]
Related features that are worth considering for future PRs:
- [ ] Local JSON support as local links.
- [x] Version number check against JSON file.

**Pull request merged**